pkg/kubelet/kubeletconfig/configsync_test.go

```diff
@@ -0,0 +1,283 @@
+	fakeEventv1 "k8s.io/client-go/kubernetes/typed/core/v1/fake"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/kubelet/kubeletconfig/checkpoint"
+	"k8s.io/kubernetes/pkg/kubelet/kubeletconfig/checkpoint/store"
+	"k8s.io/kubernetes/pkg/kubelet/kubeletconfig/status"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+	"testing"
+)
+
+var (
+	nodeMeta        = metav1.ObjectMeta{Name: "mynode", UID: "mynode-uid"}
+	configMapA1Meta = metav1.ObjectMeta{Name: "foo", Namespace: "bar", UID: "uid-a1", ResourceVersion: "12345"}
+	configMapA2Meta = metav1.ObjectMeta{Name: "foo", Namespace: "bar", UID: "uid-a2", ResourceVersion: "54321"}
+)
+
+// The configuration has not changed, no need to call os.Exit(0)
+func TestSyncConfigSourceKubeletNoRestart(t *testing.T) {
+	tmpDir, err := utilfs.DefaultFs{}.TempDir("", "store-")
+	if err != nil {
+		t.Fatalf("can't make a temp dir: %v", err)
+	}
+	defer func() {
+		_ = utilfs.DefaultFs{}.RemoveAll(tmpDir)
+	}()
+
+	testCases := map[string]struct {
+		actualNCS    *apiv1.NodeConfigSource
+		assigned     bool
+		remoteConfig apiv1.ConfigMap
+	}{
+		"source, assigned is nil": {
+			actualNCS:    nil,
+			assigned:     false,
+			remoteConfig: apiv1.ConfigMap{},
+		},
+		"source is equal to assigned": {
+			actualNCS: &apiv1.NodeConfigSource{
+				ConfigMap: &apiv1.ConfigMapNodeConfigSource{
+					Namespace:       configMapA1Meta.Namespace,
+					Name:            configMapA1Meta.Name,
+					UID:             configMapA1Meta.UID,
+					ResourceVersion: configMapA1Meta.ResourceVersion,
+				}},
+			assigned: true,
+			remoteConfig: apiv1.ConfigMap{ObjectMeta: configMapA1Meta, Data: map[string]string{
+				"--v": "0",
+			}},
+		},
+	}
+
+	cc := &Controller{
+		pendingConfigSource: make(chan bool, 1),
+		configStatus:        status.NewNodeConfigStatus(),
+		checkpointStore:     store.NewFsStore(utilfs.DefaultFs{}, tmpDir),
+	}
+	// ensure the filesystem is initialized
+	if err := cc.initializeDynamicConfigDir(); err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// fake nodeInformer
+	fakeClient := fakeclientset.NewSimpleClientset()
+	cc.nodeInformer = newSharedNodeInformer(fakeClient, nodeMeta.Name, nil, nil, nil)
+
+	// fake remoteConfigSourceInformer
+	rcs, _, err := checkpoint.NewRemoteConfigSource(&apiv1.NodeConfigSource{
+		ConfigMap: &apiv1.ConfigMapNodeConfigSource{Namespace: configMapA1Meta.Namespace, Name: configMapA1Meta.Name}})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	cc.remoteConfigSourceInformer = rcs.Informer(fakeClient,
+		cache.ResourceEventHandlerFuncs{AddFunc: nil, UpdateFunc: nil, DeleteFunc: nil})
+
+	for name, testCase := range testCases {
+		if _, err := cc.checkpointStore.Reset(); err != nil {
+			t.Fatalf("%v unexpected error %v", name, err)
+		}
+
+		node := apiv1.Node{ObjectMeta: nodeMeta, Spec: apiv1.NodeSpec{ConfigSource: testCase.actualNCS}}
+		if err := cc.nodeInformer.GetStore().Add(&node); err != nil {
```

Comment:
I think `Bootstrap `and `StartSync `have done too much, as you said, using a fake client could indeed simplify the setup.

Because I did not follow your suggestion, I need to confirm whether I still need to call `Bootstrap `and `StartSync `to initialize the Controller. :)

Reply:
I'm hoping that by using the real controller with a fake client, we can avoid a lot of the duplicate setup here (and it will help keep the test valid over time if the controller changes).

Reply:
Ok. I am going to try this.

Reply:
Hey, @mtaufen,
since k8s.io/client-go/kubernetes/fake is used, `client.CoreV1().RESTClient()` returns nil, so it is always impossible to rely on `nodeInformer.Run` of startSync here.
If we rely on startSync, we may not be able to use fakeClient... So I don't want to rely on startSync.

If you still feel that you need to rely on startSync, I will investigate.